### PR TITLE
feat: add needs_all()

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
Address https://github.com/jina-ai/jina/issues/1262
1. Add needs_all() to collect all hanging pods so far as a sugary syntax for .needs(['h1', 'h2', ...]), .needs_all() can be used at any part of the Flow.
2. add test_flow_needs_all() in unit test test_flow.py

### CODE REVIEW

File: tests/unit/flow/test_flow.py [hunk 1]
```diff
@@ -559,3 +559,17 @@ def test_flow_arbitrary_needs():
...
+         .add(name='p2', needs='gateway')
+         .add(name='p3', needs='gateway')
+         .needs(needs=['p1', 'p2'], name='r1')
+         .needs_all(name='r2')
+         )
+
+    all_needs = {v for p in f._pod_nodes.values() for v in p.needs}
+    all_names = {p for p in f._pod_nodes.keys()}
+
+    assert list(all_names.difference(all_needs)) == ['r2']
```
Comment (nan-wang) [inline tests/unit/flow/test_flow.py:575]: This test is not so explictie. Check the `needs` of `r2` might be easier to understand.

**Review by nan-wang — CHANGES_REQUESTED** [review]
Minor comments.

File: jina/flow/__init__.py [hunk 2]
```diff
@@ -251,6 +251,19 @@ def needs(self, needs: Union[Tuple[str], List[str]],
...
+    def needs_all(self, copy_flow: bool = True,
+                  name: str = 'joiner', *args, **kwargs) -> 'Flow':
+        """
+        Collect all hanging Pod so far and add a blocker to the flow, wait until all handing peas completed.
+        :param copy_flow: when set to true, then always copy the current flow and do the modification on top of it then return, otherwise, do in-line modification
+        :param name: the name of this joiner, by default is ``joiner``
+        :return: the modified flow
+        """
+        op_flow = copy.deepcopy(self) if copy_flow else self
+        needs = _hanging_pods(op_flow)
```
Comment (hanxiao) [inline jina/flow/__init__.py:263]: wont work

File: jina/flow/__init__.py [hunk 3]
```diff
@@ -251,6 +251,19 @@ def needs(self, needs: Union[Tuple[str], List[str]],
...
+
+        return self.needs(needs=needs, name=name, *args, **kwargs)
```
Comment (maximilianwerk) [inline jina/flow/__init__.py:265]: This won't work, if there is only one hanging pod. Thus, you can either add a flag `accept_one_pod` to the `needs` function in order to reuse `needs` or directly call `add` here.

**Review by maximilianwerk — CHANGES_REQUESTED** [review]
Thanks for submitting you PR. Besides the comments below, could you also add more tests for:

- run with some some random data, e.g. `f.index_ndarray(np.random.random([10,10]))`
- `needs_all` called in the middle of the flow and not just at the very end.

Cheers!

File: tests/unit/flow/test_flow.py (hunk 1 above)
Comment (Yongxuanzhang) [reply tests/unit/flow/test_flow.py]: Sure! It has been updated accordingly, please let me know if that works.

File: jina/flow/__init__.py (hunk 3 above)
Comment (Yongxuanzhang) [reply jina/flow/__init__.py:265]: I have updated the code to fix these issues accordingly and also added required unit tests, please let me know if that works. Thanks!

Comment (maximilianwerk) [reply jina/flow/__init__.py:265]: Looks great so far. One tiny cleanup: Could you switch the order of variables in the calling of `add` and `needs` to be the same? That'll improve readability.

Comment (Yongxuanzhang) [reply jina/flow/__init__.py]: Sure! Thank you for the advice! 
I have made the change to rearrange them and keep it consistent. 
I have also thought a little about it before and the reason I arranged the order of variables this way is that I wanted to keep it consistent with current conventions. (e.g. In the unit test test_flow_arbitrary_needs(), .add() is called as (name, needs) and .needs is called as (needs, name).

**Review by hanxiao — APPROVED** [review]
looking good! thank you!

**Comment by hanxiao** [comment]
recheckcla

File: tests/unit/flow/test_flow.py [hunk 4]
```diff
@@ -559,3 +560,31 @@ def test_flow_arbitrary_needs():
...
+
+    f = (Flow().add(name='p1', needs='gateway')
+         .add(name='p2', needs='gateway')
+         .add(name='p3', needs='gateway')
+         .needs(needs=['p1', 'p2'], name='r1')
+         .needs_all(name='r2'))
+    assert f._pod_nodes['r2'].needs == {'p3', 'r1'}
+
+    with f:
+        f.index_ndarray(np.random.random([10, 10]))
```
Comment (nan-wang) [inline tests/unit/flow/test_flow.py:578]:
```suggestion

```

File: tests/unit/flow/test_flow.py [hunk 5]
```diff
@@ -559,3 +560,31 @@ def test_flow_arbitrary_needs():
...
+         .add(name='p2', needs='gateway')
+         .add(name='p3', needs='gateway')
+         .needs(needs=['p1', 'p2'], name='r1')
+         .needs_all(name='r2')
+         .add(name='p4', needs='r2'))
+    assert f._pod_nodes['r2'].needs == {'p3', 'r1'}
+    assert f._pod_nodes['p4'].needs == {'r2'}
+
+    with f:
+        f.index_ndarray(np.random.random([10, 10]))
```
Comment (nan-wang) [inline tests/unit/flow/test_flow.py:590]: It is not necessary to run the indexing for this test. 
```suggestion

```

**Review by nan-wang — CHANGES_REQUESTED** [review]
minor changes

File: tests/unit/flow/test_flow.py (hunk 4 above)
Comment (Yongxuanzhang) [reply tests/unit/flow/test_flow.py:578]: Thanks for your comment!
Those were added to address  (https://github.com/jina-ai/jina/pull/1288#pullrequestreview-531216135)  requested  by @maximilianwerk or did I misunderstood the requirement? Feel free to let me know whether  keep them or replace those tests with better ones. Thanks!

Comment (maximilianwerk) [reply tests/unit/flow/test_flow.py:578]: I had a short chat with @hanxiao and there we agreed, that a short test with data flowing through the flow would be beneficial in order to test the empty pods.

Comment (hanxiao) [reply tests/unit/flow/test_flow.py:578]: yes. i think its all good, merge shall we?

**Review by maximilianwerk — APPROVED** [review]

**Comment by Yongxuanzhang** [comment]
I have read the CLA Document and I hereby sign the CLA

**Pull request merged**